Add action to reset Schwarz subdomain solver

nilsvu opened the pull request:
## Proposed changes

This action will be called in nonlinear solver iterations to reset caches of the Schwarz subdomain solver. The reset can be disabled to potentially speed up solves.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add action to reset Schwarz subdomain solver

wthrowe commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Actions/ResetSubdomainSolver.hpp:
```diff
@@ -0,0 +1,95 @@
+                   ::Verbosity::Debug)) {
+        Parallel::printf("%s " + Options::name<OptionsGroup>() +
+                             ": Reset subdomain solver\n",
+                         element_id);
```
`"%s %s:  Reset subdomain solver\n", element_id, Options::name<OptionsGroup>()` (So it doesn't segfault if `Options::name<OptionsGroup>()` happens to have a % in it.)

wthrowe commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Actions/ResetSubdomainSolver.hpp:
```diff
@@ -0,0 +1,95 @@
+ * \brief Reset the subdomain solver, clearing its caches related to the linear
+ * operator it has solved so far.
+ *
+ * Invoke this action when the linear operator changed. For example, an operator
```
Either "changes" or "has changed".

nilsvu committed:
Pass linear solver name to printf as argument

This is safer than concatenating strings since it prevents
problems when the name contains '%' symbols.

nilsvu committed:
Make linear solver logging more consistent

nilsvu commented:
@wthrowe I posted a fixup to address your changes, and also added a commit that propagates the change for your first comment to the rest of the actions. While doing that I noticed a bunch of inconsistencies that I fixed in another commit (I could push that to a separate PR if you prefer).

wthrowe commented on src/NumericalAlgorithms/Convergence/HasConverged.cpp:
```diff
@@ -84,15 +85,16 @@ std::ostream& operator<<(std::ostream& os,
                << "AbsoluteResidual - The residual magnitude has decreased to "
-               << has_converged.criteria_.absolute_residual << " or below ("
-               << has_converged.residual_magnitude_ << ").";
+               << get_output(has_converged.criteria_.absolute_residual)
```
Why are you running these through `get_output`?

nilsvu replied:
To print them at a reasonable precision without having to do things like `<< std::setprecision`

wthrowe replied:
But `get_output` just prints at the default precision.  If anything, I would expect any stream passed in to be at a higher precision than that.

nilsvu replied:
Default precision is good. Without the `get_output` this prints 16(ish) digits.

wthrowe replied:
I see that we do indeed set the precision to something ridiculous. Looks good then.

wthrowe commented on src/ParallelAlgorithms/LinearSolver/AsynchronousSolvers/ElementActions.hpp:
```diff
@@ -81,18 +81,15 @@ void contribute_to_residual_observation(
-          "residual: %e\n",
-          Options::name<OptionsGroup>(), get_output(array_index),
-          sqrt(residual_magnitude_square));
+      Parallel::printf("%s %s initialized with local residual: %e\n",
```
Do you think it would be a good idea to keep "Linear solver" in the message so people looking at the log know what part of the code the message is from?

nilsvu replied:
Hmm, you mean print them like `[ElementId] Linear solver GMRES(IterationId): ...`? I think it's better to make the name of the solver reasonably descriptive to identify where the messages are coming from. E.g. I'll name them `SchwarzSmoother` and `Multigrid`.

wthrowe replied:
I guess my real question is: Is everyone who runs the executable expected to know that "Multigrid" is a linear solver?  If yes, then this is fine.

nilsvu replied:
I think when they read "multigrid" then they can go look up the multigrid code, which happens to be a linear solver. Sound ok?

Edit: What I mean is: There's going to be files in the repo named 'Multigrid', as there are currently already files named 'Schwarz'.

wthrowe replied:
I'm more worried about the other direction:  I want to read the log entries about the linear solver, but I don't know which ones those are.  I may be overestimating how much of a problem that will be, though, so let's wait and see how things look when we have everything fully running.

wthrowe commented on src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ElementActions.hpp:
```diff
@@ -323,10 +324,11 @@ struct PerformStep {
           db::get<Convergence::Tags::IterationId<OptionsGroup>>(box),
-          db::get<NonlinearSolver::Tags::StepLength<OptionsGroup>>(box));
+          get_output(
+              db::get<NonlinearSolver::Tags::StepLength<OptionsGroup>>(box)));
```
I'm pretty sure this formats like "%g", maybe with a precision indicated.

nilsvu replied:
Thanks, pushed another fixup

wthrowe reviewed (changes requested):
The fixup can be squashed.

I had some comments on the new commit.  We can either resolve them here or as a separate PR, your choice.

wthrowe reviewed (commented):
Squash.

nilsvu commented:
Rebased and squashed, thanks @wthrowe. @moxcodes could you do the second review for this?

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Actions/ResetSubdomainSolver.hpp:
```diff
@@ -0,0 +1,94 @@
+      db::mutate<
+          LinearSolver::Schwarz::Tags::SubdomainSolverBase<OptionsGroup>>(
+          make_not_null(&box), [](const auto subdomain_solver) noexcept {
+            // Dereference the gsl::not_null pointer, and then the
```
Can the type logic be clarified by just taking the argument as a `not_null` of a `unique_ptr` to the abstract base class in question instead of `auto`? Or is that base class not known here?

nilsvu replied:
No the base class is not known here

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Actions/Test_ResetSubdomainSolver.cpp:
```diff
@@ -0,0 +1,90 @@
+  enum class Phase { Initialization, Testing, Exit };
+};
+
+void test_reset_subdomain_solver(const bool skip_resets) {
```
`noexcept`

nilsvu replied:
I don't think we have been strict about this in tests at all 😄 Added it though.

moxcodes reviewed (commented):
Looks good, just a couple of minor things.

nilsvu commented:
Thanks for reviewing @moxcodes and @wthrowe!

nilsvu commented:
Rebased on #2996 to fix CI failure
